distributed/core.py

```diff
@@ -238,7 +238,8 @@ def set_thread_ident():
 
         self.__stopped = False
 
-        super().__init__(**kwargs)
+        if kwargs:
+            logger.warning("Provided unused arguments %s", list(kwargs.keys()))
```

Comment:
Hmm I think the call to `super().__init__` was intentional -- see https://github.com/dask/distributed/pull/4365/files/2212edc11b424c17f2f656eb73b2ca206742705a#r557630488 for an earlier discussion on this point

Reply:
multiple inheritance is a wonderful monstrosity :D

```python
In [1]: class A:
   ...:     def __init__(self, a, **kwargs):
   ...:         print(f"A: {kwargs}")
   ...:

In [2]: class B(A):
   ...:     def __init__(self, b, **kwargs):
   ...:         print(f"B: {kwargs}")
   ...:         super().__init__(**kwargs)
   ...:

In [3]: class C(A):
   ...:     def __init__(self, c, **kwargs):
   ...:         print(f"C: {kwargs}")
   ...:         super().__init__(**kwargs)
   ...:

In [4]: class D(B, C):
   ...:     def __init__(self, d, **kwargs):
   ...:         print(f"D: {kwargs}")
   ...:         super().__init__(**kwargs)
   ...:

In [5]: D.mro()
Out[5]: [__main__.D, __main__.B, __main__.C, __main__.A, object]

In [6]: D(a="a", b="b", c="c", d="d")
D: {'a': 'a', 'b': 'b', 'c': 'c'}
B: {'a': 'a', 'c': 'c'}
C: {'a': 'a'}
A: {}
```

kwargs will trickle down the mro. therefore we need to allow kwargs everwhere. This is true for all classes but the root. the root cannot call super with kwargs since object doesn't allow arguments. afaik, the init on object is a no-op

Reply:
We could consider adding a test like this (untested):
```python
def test_server_inheritance():
    subclasses = Server.__subclasses__()
    for subcls in subclasses:
        assert subcls.__mro__[-2:] == [Server, object], f"Server must always be the last class in the MRO. {subcls} has MRO: {subcls.__mro__}"
```
Obviously this would depend on all Server subclasses being imported at the time the test runs.

Alternatively, if we wanted to over-engineer it a bit, we could use [`__init_subclass__`](https://docs.python.org/3/reference/datamodel.html#customizing-class-creation) (added in 3.6) to enforce this. I've done something like this before [here](https://github.com/descarteslabs/descarteslabs-python/blob/master/descarteslabs/workflows/types/widget/widget.py#L42-L57). Not sure how it would play with Cython though?